[MacOS][Pluggable Device] Fixes for macos pluggable device.

kulinseth opened the pull request:
1. Disable the Eigen max align check as we use custom device memory
   allocator in Metal plugin
2. Disable list kernels on Pluggable devices as Device memory on MacOS
   pluggable device doesn't allow us to index at a offset from the base
   MTLBuufer location.
3. Disable XLA compilation path as its not supported in pluggable
   architecture
cc. @penpornk , @learning-to-play , @nitins17

penpornk commented on tensorflow/core/framework/tensor.h:
```diff
@@ -35,6 +35,12 @@ limitations under the License.
 #include "tensorflow/core/platform/mem.h"
 #include "tensorflow/core/platform/types.h"
 
+#if !defined(PLUGGABLE_DEVICE_SUPPORTED) &&                                \
```
Do you mean
```suggestion
#if !defined(PLUGGABLE_DEVICE_SUPPORTED_MACOS) &&                                \
```
?

kulinseth replied:
Yes, as discussed, I have created a separate Macro for this.

penpornk commented on tensorflow/core/framework/tensor.h:
```diff
@@ -35,6 +35,12 @@ limitations under the License.
 #include "tensorflow/core/platform/types.h"
 
+#if !defined(PLUGGABLE_DEVICE_SUPPORTED) &&                                \
+    (__x86_64__ || __i386__ || defined(__APPLE__)) && !defined(ANDROID) && \
+    !defined(__ANDROID__) && !TARGET_OS_IOS
```
These conditions are also for non-MacOS (any Linux on x86_64 and i386 will also satisfy the condition).
https://github.com/tensorflow/tensorflow/blob/5f7dcec3fabfa7b119790f87db20775088f0e2b7/tensorflow/core/kernels/list_kernels.h#L42-L49

To make it MacOS-specific, I think we only need
```suggestion
    defined(__APPLE__)) && !TARGET_OS_IOS
```

If you think this should apply to all PluggableDevice, then maybe we should rename it to something more generic.

kulinseth replied:
Done.

penpornk reviewed (commented):
Thank you for the PR!

@Jianhui-Li @jzhoulon @yiqianglee @PatriceVignola Would you like your plug-ins to disable Eigen max align check as well?

kulinseth committed:
[MacOS][Pluggable Device] Fixes for macos pluggable device.

1. Disable the Eigen max align check as we use custom device memory
   allocator in Metal plugin
2. Disable list kernels on Pluggable devices as Device memory on MacOS
   pluggable device doesn't allow us to index at a offset from the base
   MTLBuufer location.
3. Disable XLA compilation path as its not supported in pluggable
   architecture

Jianhui-Li commented:
> Thank you for the PR!
> 
> @Jianhui-Li @jzhoulon @yiqianglee @PatriceVignola Would you like your plug-ins to disable Eigen max align check as well?

I don't think we need this check for Intel Plugin. It seems to me an implementation issue specific to Metal plugin.  Even the device allocator doesnt' support alignment, the plugin could enforce alignment on top of it.

penpornk commented on tensorflow/core/framework/tensor.h:
```diff
@@ -35,6 +35,12 @@ limitations under the License.
 #include "tensorflow/core/platform/types.h"
 
+#if !defined(PLUGGABLE_DEVICE_SUPPORTED_MACOS) &&  \
+    defined(__APPLE__) && !defined(ANDROID) &&     \
+    !defined(__ANDROID__) && !TARGET_OS_IOS
```
Nit: `!defined(ANDROID)` appeared twice.
```suggestion
    defined(__APPLE__) && !defined(ANDROID) && !TARGET_OS_IOS
```

penpornk replied:
@kulinseth  Sorry I misread ANDROID and __ANDROID__. I thought they were the same (just __ANDROID__ twice). I'll add this back internally!

penpornk reviewed (changes requested):
@Jianhui-Li Thank you for the quick reply!

kulinseth committed:
Update tensorflow/core/framework/tensor.h

Co-authored-by: Penporn Koanantakool <38085909+penpornk@users.noreply.github.com>

penpornk reviewed (approved):
Thank you again!
